python_scripts/parameter_tuning_nested.py

```diff
@@ -69,56 +66,155 @@
+# to obtain the mean and standard deviation of the scores. We will use the
+# `cross_validate` function to evaluate our model.
 
 # %%
 from sklearn.model_selection import cross_validate
+
+cv_results = cross_validate(model, data, target, cv=5)
+cv_results = pd.DataFrame(cv_results)
+cv_results
+
+# %%
+# The cross-validation scores are coming from a 5-fold cross-validation. So
+# we can compute the mean and standard deviation of the generalization score.
+
+# %%
+print(
+    "Generalization score without hyperparameters tuning:\n"
+    f"{cv_results['test_score'].mean():.3f} +/- {cv_results['test_score'].std():.3f}"
+)
+
+# %% [markdown]
+# Now, we will present how to evaluate the model with hyperparameter tuning,
+# where an extra step is required to select the best set of parameters.
+#
+# ### With hyperparameter tuning
+#
+# As we shown in the previous notebook, one can use a search strategy that uses
+# cross-validation to find the best set of parameters. Here, we will use a
+# grid-search strategy and reproduce the steps done in the previous notebook.
+#
+# First, we have to embed our model into a grid-search and specify the
+# parameters and the parameter values that we want to explore.
+
+# %%
 from sklearn.model_selection import GridSearchCV
 
 param_grid = {
-    'classifier__learning_rate': (0.05, 0.1),
-    'classifier__max_leaf_nodes': (30, 40)}
-model_grid_search = GridSearchCV(model, param_grid=param_grid,
-                                 n_jobs=2, cv=2)
+    'classifier__learning_rate': (0.05, 0.5),
+    'classifier__max_leaf_nodes': (10, 30)}
+model_grid_search = GridSearchCV(
+    model, param_grid=param_grid, n_jobs=2, cv=2
+)
+model_grid_search.fit(data, target)
 
-cv_results = cross_validate(
-    model_grid_search, data, target, cv=3, return_estimator=True)
+# %% [markdown]
+# As previously saw, when calling the `fit` method, the model embedded in the
+# grid-search will be trained with every possible combinations of parameters
+# resulting from the parameter grid. The best combination was selected by
+# keeping the combination of the best mean accuracy score.
+
+# %%
+cv_results = pd.DataFrame(model_grid_search.cv_results_)
+cv_results[[
+    "param_classifier__learning_rate",
+    "param_classifier__max_leaf_nodes",
+    "mean_test_score",
+    "rank_test_score"
+]]
+
+# %%
+model_grid_search.best_params_
 
 # %% [markdown]
-# Running the above cross-validation will give us an estimate of the
-# testing score.
+# One important bottleneck here concerns the generalization performance.
+# Indeed, the mean and standard deviation of the generalization score provided
+# by the cross-validation in the grid-search is potentially not a good estimate
+# of the generalization performance of the model refitted with the best
+# parameters (i.e. the model at hand when using `model_grid_search.predict`).
+#
+# The reason is that the model was refitted on the full dataset given during
```

Comment:
Maybe a matter of taste, but I think the argument in the cross_validation_nested.py notebook is clearer:

>  we should not forget that we used this score to pick-up the best model. It means that we used knowledge from the test set (i.e. test score) to decide our model’s training parameter.

Reply:
I think both points are interesting (using the best mean internal CV score to select the model to refit and the fact that we refit on the full data after selecting the best model).

However I think it confusing to mention "the test set" in the context of a cross-validation since there are several inner test sets.